src/frontend/src/planner/select.rs

```diff
@@ -101,56 +101,39 @@ impl Planner {
             Condition::with_expr(where_clause)
                 .group_by::<_, 3>(|expr| match expr {
                     ExprImpl::Subquery(_) => 0,
-                    ExprImpl::FunctionCall(func_call)
-                        if func_call.get_expr_type() == ExprType::Not
-                            && matches!(func_call.inputs()[0], ExprImpl::Subquery(_)) =>
-                    {
-                        1
-                    }
+                    ExprImpl::FunctionCall(func_call) => match func_call.get_expr_type() {
+                        ExprType::In if matches!(func_call.inputs()[1], ExprImpl::Subquery(_)) => 0,
+                        ExprType::Not => {
+                            let expr = &func_call.inputs()[0];
+                            match expr {
+                                ExprImpl::Subquery(_) => 1,
+                                ExprImpl::FunctionCall(func)
+                                    if func.get_expr_type() == ExprType::In
+                                        && matches!(func.inputs()[1], ExprImpl::Subquery(_)) =>
```

Comment:
Maybe we can simply the matching by updating the definition of `expr::Subquery`:

Option A:
```
pub struct Subquery {
    pub query: BoundQuery,
    pub kind: SubqueryKind,
    pub negated: bool, // only meaningful for Existential or SetComparison, maybe Option<bool>
    pub operand: Option<ExprImpl>,  // only meaningful for SetComparison
}
```

Option B:
```
pub enum SubqueryKind {
    Scalar,
    Existential(bool),
    SetComparison(bool, ExprImpl),
}
```

This way we avoid packing and unpacking with `ExprType::In` or `ExprType::Not`.

Note that this is still not enough to represent `a >= ALL(...)` or `a <> SOME(...)` yet. The comparison operator also needs to be stored, as well as differentiating `ALL` from `SOME`.

Reply:
+1

Reply:
An additional reason is that calling `q` in `e in q` "SetComparison" looks a little strange.

Reply:
IMO, refining the definition of `Subquery` or `SubqueryKind` deserves a new pr. As for these two options, extending `SubqueryKind` looks better for me, but it still causes problems when do the matching because we need to extract `SubqueryKind` from `Subquery` now. In conclusion, defining a good struct for Subquery is not that simple.